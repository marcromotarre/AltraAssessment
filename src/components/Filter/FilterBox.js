import React from 'react';
import styled from 'styled-components';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import FiltersApplied from './FiltersApplied';
import FilterSetter from './FilterSetter';
import FilterSVG from '../../svg/filter.svg';

export const FilterContainer = styled.section`
  background: #f7f8fd;
  width: 100%;
  display: grid;
  grid-template-columns: 40px auto;
`;

export const FilterButtonContainer = styled.div`
  button {
    padding-left: 10px;
    padding-top: 10px;
    justify-self: center;
    align-self: start;
  }
`;

const FilterBox = ({
  filteredUsersNumber,
  filteringMode,
  setFilteringMode,
  isAnyFilterApplied,
}) => {
  const filters = useSelector((state) => state.filters);
  const handleOnClickFilterIcon = () => setFilteringMode(!filteringMode);

  return (
    <FilterContainer>
      <FilterButtonContainer>
        {(isAnyFilterApplied || filteringMode) && (
          <button onClick={handleOnClickFilterIcon}>
            <FilterSVG width={21.5} />
          </button>
        )}
      </FilterButtonContainer>
      {isAnyFilterApplied && !filteringMode && <FiltersApplied {...filters} />}
      {filteringMode && (
        <FilterSetter filteredUsersNumber={filteredUsersNumber} />
      )}
    </FilterContainer>
  );
};

FilterBox.propTypes = {
  filteredUsersNumber: PropTypes.number.isRequired,
  filteringMode: PropTypes.bool,
  setFilteringMode: PropTypes.func,
  isAnyFilterApplied: PropTypes.bool,
};

FilterBox.defaultProps = {
  filteringMode: false,
  setFilteringMode: () => {},
  isAnyFilterApplied: false,
};

export default FilterBox;
