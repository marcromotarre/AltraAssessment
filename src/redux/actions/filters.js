import * as types from './../constants';

export const setAgeFilter = ({ filterData }) => ({
  type: types.SET_AGE_FILTER,
  payload: {
    filterData,
  },
});

export const setAgeMin = ({ min }) => ({
  type: types.SET_AGE_MIN,
  payload: {
    min,
  },
});

export const setAgeMax = ({ max }) => ({
  type: types.SET_AGE_MAX,
  payload: {
    max,
  },
});

export const setAgeRange = ({ range }) => ({
  type: types.SET_AGE_RANGE,
  payload: {
    range,
  },
});

export const setHeightFilter = ({ filterData }) => ({
  type: types.SET_HEIGHT_FILTER,
  payload: {
    filterData,
  },
});

export const setHeightMin = ({ min }) => ({
  type: types.SET_HEIGHT_MIN,
  payload: {
    min,
  },
});

export const setHeightMax = ({ max }) => ({
  type: types.SET_HEIGHT_MAX,
  payload: {
    max,
  },
});

export const setHeightRange = ({ range }) => ({
  type: types.SET_HEIGHT_RANGE,
  payload: {
    range,
  },
});

export const setWeightFilter = ({ filterData }) => ({
  type: types.SET_WEIGHT_FILTER,
  payload: {
    filterData,
  },
});

export const setWeightMin = ({ min }) => ({
  type: types.SET_WEIGHT_MIN,
  payload: {
    min,
  },
});

export const setWeightMax = ({ max }) => ({
  type: types.SET_WEIGHT_MAX,
  payload: {
    max,
  },
});

export const setWeightRange = ({ range }) => ({
  type: types.SET_WEIGHT_RANGE,
  payload: {
    range,
  },
});

export const addHairColorFilter = ({ color }) => ({
  type: types.ADD_HAIR_COLOR_FILTER,
  payload: {
    color,
  },
});

export const deleteHairColorFilter = ({ color }) => ({
  type: types.DELETE_HAIR_COLOR_FILTER,
  payload: {
    color,
  },
});

export const clearHairColorsFilters = () => ({
  type: types.CLEAR_HAIR_COLORS_FILTER,
});
