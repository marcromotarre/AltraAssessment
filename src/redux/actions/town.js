import * as types from './../constants';

export function fetchTownPending() {
  return {
    type: types.FETCH_TOWN_PENDING,
  };
}

export function fetchTownSuccess(response) {
  return {
    type: types.FETCH_TOWN_SUCCESS,
    payload: {
      response,
    },
  };
}

export function fetchTownError(error) {
  return {
    type: types.FETCH_TOWN_ERROR,
    payload: {
      error,
    },
  };
}
