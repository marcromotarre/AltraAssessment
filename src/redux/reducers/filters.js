import * as types from './../constants';

export const initialState = {
  age: {
    min: '',
    max: '',
    range: true,
  },
  weight: {
    min: '',
    max: '',
    range: true,
  },
  height: {
    min: '',
    max: '',
    range: true,
  },
  hair_colors: [],
};

export function reducer(state = initialState, { type, payload }) {
  switch (type) {
    case types.SET_AGE_FILTER:
      return {
        ...state,
        age: payload.filterData,
      };
    case types.SET_AGE_MIN:
      return {
        ...state,
        age: { ...state.age, min: payload.min },
      };
    case types.SET_AGE_MAX:
      return {
        ...state,
        age: { ...state.age, max: payload.max },
      };
    case types.SET_AGE_RANGE:
      return {
        ...state,
        age: { ...state.age, range: payload.range },
      };
    case types.SET_HEIGHT_FILTER:
      return {
        ...state,
        height: payload.filterData,
      };
    case types.SET_HEIGHT_MIN:
      return {
        ...state,
        height: { ...state.height, min: payload.min },
      };
    case types.SET_HEIGHT_MAX:
      return {
        ...state,
        height: { ...state.height, max: payload.max },
      };
    case types.SET_HEIGHT_RANGE:
      return {
        ...state,
        height: { ...state.height, range: payload.range },
      };
    case types.SET_WEIGHT_FILTER:
      return {
        ...state,
        weight: payload.filterData,
      };
    case types.SET_WEIGHT_MIN:
      return {
        ...state,
        weight: { ...state.weight, min: payload.min },
      };
    case types.SET_WEIGHT_MAX:
      return {
        ...state,
        weight: { ...state.weight, max: payload.max },
      };
    case types.SET_WEIGHT_RANGE:
      return {
        ...state,
        weight: { ...state.weight, range: payload.range },
      };
    case types.ADD_HAIR_COLOR_FILTER:
      return {
        ...state,
        hair_colors: !state.hair_colors.includes(payload.color)
          ? [...state.hair_colors, payload.color]
          : [...state.hair_colors],
      };

    case types.DELETE_HAIR_COLOR_FILTER:
      return {
        ...state,
        hair_colors: state.hair_colors.filter(
          (color) => color !== payload.color
        ),
      };

    case types.CLEAR_HAIR_COLORS_FILTER:
      return {
        ...state,
        hair_colors: [],
      };

    default:
      return {
        ...state,
      };
  }
}
