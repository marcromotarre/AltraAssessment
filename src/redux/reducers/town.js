import * as types from './../constants';
import {
  getTownsNames,
  getUsersByTown,
  getUniqueHairColorsByTown,
} from '../../utils/town';

export const initialState = {
  pending: false,
  townName: '',
  users: [],
  uniqueHairColors: [],
  error: null,
};

export function reducer(state = initialState, { type, payload }) {
  switch (type) {
    case types.FETCH_TOWN_PENDING:
      return {
        ...state,
        pending: true,
      };
    case types.FETCH_TOWN_SUCCESS:
      const townName = getTownsNames(payload.response)[0];
      return {
        ...state,
        pending: false,
        townName,
        users: getUsersByTown(payload.response, townName),
        uniqueHairColors: getUniqueHairColorsByTown(payload.response, townName),
      };
    case types.FETCH_TOWN_ERROR:
      return {
        ...state,
        pending: false,
        error: payload.error,
      };

    default:
      return {
        ...state,
      };
  }
}
