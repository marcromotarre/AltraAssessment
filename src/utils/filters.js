export const isMinMaxFilterApplied = (filter) =>
  filter.min === '' && filter.max === '' ? false : true;

export const isArrayFilterApplied = (filter) =>
  filter.length > 0 ? true : false;

export const getFilterMinMaxString = (data) => {
  if (data.range && data.min !== '' && data.max !== '') {
    return `${data.min} - ${data.max}`;
  } else if (data.min !== '') {
    return `${data.min}`;
  } else if (data.max !== '') {
    return `${data.max}`;
  } else {
    return '';
  }
};

export const isAgeOrWeightOrHeightFilterApplied = ({ age, height, weight }) => {
  return isMinMaxFilterApplied(age) ||
    isMinMaxFilterApplied(height) ||
    isMinMaxFilterApplied(weight)
    ? true
    : false;
};

export const isAnyFilterApplied = ({ age, height, weight, hair_colors }) => {
  return isAgeOrWeightOrHeightFilterApplied({ age, height, weight }) ||
    isArrayFilterApplied(hair_colors)
    ? true
    : false;
};

export const getUsersByName = (users, name) => {
  return users.filter((user) =>
    user.name.toLowerCase().includes(name.toLowerCase())
  );
};

export const getUsersByProfession = (users, profession) => {
  return users.filter((user) => user.professions.includes(profession));
};

export const getUniqueHairColors = (users) => {
  return users
    .map((user) => user.hair_color)
    .filter((value, index, self) => self.indexOf(value) === index);
};

export const getUsersByMinMaxFilter = (users, type, filter) => {
  if (!isMinMaxFilterApplied(filter)) {
    return [...users];
  } else {
    if (filter.min !== '' && filter.max !== '' && filter.range) {
      return users.filter((user) => {
        return (
          user[type] >= parseFloat(filter.min) &&
          user[type] <= parseFloat(filter.max)
        );
      });
    } else if (
      (filter.min !== '' && filter.max !== '' && !filter.range) ||
      filter.min !== ''
    ) {
      return users.filter((user) => {
        return Math.round(user[type]) === parseFloat(filter.min);
      });
    } else {
      return users.filter((user) => {
        return Math.round(user[type]) === parseFloat(filter.max);
      });
    }
  }
};

export const getUsersByArrayFilter = (users, type, filter) => {
  if (!isArrayFilterApplied(filter)) {
    return [...users];
  } else {
    return users.filter((user) => {
      return filter.includes(user[type]);
    });
  }
};

export const getUsersByFilters = ({
  users = [],
  age = { min: '', max: '', range: false },
  height = { min: '', max: '', range: false },
  weight = { min: '', max: '', range: false },
  hair_colors = [],
}) => {
  return getUsersByArrayFilter(
    getUsersByMinMaxFilter(
      getUsersByMinMaxFilter(
        getUsersByMinMaxFilter(users, 'age', age),
        'weight',
        weight
      ),
      'height',
      height
    ),
    'hair_color',
    hair_colors
  );
};
