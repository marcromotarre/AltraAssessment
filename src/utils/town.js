import { getUniqueHairColors } from './filters';

export const getTownsNames = (data) => Object.keys(data);

export const getUsersByTown = (data, townName) => data[townName];

export const getUniqueHairColorsByTown = (data, townName) =>
  getUniqueHairColors(data[townName]);
