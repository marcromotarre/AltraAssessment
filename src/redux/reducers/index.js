import { combineReducers } from 'redux';
import { reducer as town } from './town';
import { reducer as filters } from './filters';

export default combineReducers({
  town,
  filters,
});
