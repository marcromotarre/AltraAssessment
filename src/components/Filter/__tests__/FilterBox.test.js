import React from 'react';
import { fireEvent } from '@testing-library/react';
import FilterBox from '../FilterBox';

import { initialState as filtersInitialState } from './../../../redux/reducers/filters';
import { initialState as townInitialState } from './../../../redux/reducers/town';
import { renderRedux } from '../../../utils/test';

/**
 * this component manages Filter mode (Filter Setter o FilterApply )
 */

const props = {
  filteredUsersNumber: 1000,
};

const initialState = {
  filters: filtersInitialState,
  town: townInitialState,
};

describe('FilterBox Component tests', () => {
  it('FilterBox Snapshot', () => {
    expect(
      renderRedux(<FilterBox {...props} />, {
        initialState: initialState,
      }).baseElement
    ).toMatchSnapshot();
  });

  it('FilterBox when !isAnyFilterApplied && !filteringMode show not paint FiltersApplied nor FilterSetter', () => {
    const { queryByTestId } = renderRedux(
      <FilterBox
        filteredUsersNumber={1000}
        isAnyFilterApplied={false}
        filteringMode={false}
      />,
      {
        initialState: initialState,
      }
    );

    expect(queryByTestId('FiltersApplied')).toBeFalsy();
    expect(queryByTestId('FilterSetter')).toBeFalsy();
  });

  it('FilterBox when !isAnyFilterApplied && filteringMode should paint Filters Applied', () => {
    const { queryByTestId } = renderRedux(
      <FilterBox
        filteredUsersNumber={1000}
        isAnyFilterApplied={false}
        filteringMode={true}
      />,
      {
        initialState: initialState,
      }
    );
    expect(queryByTestId('FiltersApplied')).toBeFalsy();
    expect(queryByTestId('FilterSetter')).toBeTruthy();
  });

  it('FilterBox when isAnyFilterApplied && !filteringMode should paint Filters Applied', () => {
    const { queryByTestId } = renderRedux(
      <FilterBox
        filteredUsersNumber={1000}
        isAnyFilterApplied={true}
        filteringMode={false}
      />,
      {
        initialState: initialState,
      }
    );
    expect(queryByTestId('FiltersApplied')).toBeTruthy();
    expect(queryByTestId('FilterSetter')).toBeFalsy();
  });

  it('FilterBox when isAnyFilterApplied && filteringMode should paint Filters Applied', () => {
    const { queryByTestId } = renderRedux(
      <FilterBox
        filteredUsersNumber={1000}
        isAnyFilterApplied={true}
        filteringMode={false}
      />,
      {
        initialState: initialState,
      }
    );
    expect(queryByTestId('FiltersApplied')).toBeTruthy();
    expect(queryByTestId('FilterSetter')).toBeFalsy();
  });

  it('FilterBox when press filter icon button should call prop setFilteringMode', () => {
    const setFilteringMode = jest.fn();
    const { container } = renderRedux(
      <FilterBox
        filteredUsersNumber={1000}
        isAnyFilterApplied={true}
        filteringMode={false}
        setFilteringMode={setFilteringMode}
      />,
      {
        initialState: initialState,
      }
    );
    expect(setFilteringMode).toHaveBeenCalledTimes(0);
    const button = container.querySelector('button');
    fireEvent.click(button);
    expect(setFilteringMode).toHaveBeenCalledTimes(1);
  });
});
