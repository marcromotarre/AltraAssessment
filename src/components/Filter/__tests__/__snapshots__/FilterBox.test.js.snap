// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBox Component tests FilterBox Snapshot 1`] = `
<body>
  <div>
    <section
      class="sc-fzoLag Apptz"
    >
      <div
        class="sc-fzoXzr lixDwH"
      />
    </section>
  </div>
</body>
`;
